File: алгоритмы перебора/Задача коммивояжера. Перебор с отсечениями/Задача коммивояжера. Перебор с отсечениями.py
def count(p):
    global a
    sum = 0
    for i in range(0, len(p)):
        if (i)== 0:
            sum+=a[0][p[i]]
            #print ("из первого города во второй - " + str(a[0][p[i]]))
        elif (i) == len(p):
            sum+=a[p[i]][0]
            #print ("из предпоследнего города в последний - " + str(a[p[i]][0]))
        else:
            sum+=a[p[i-1]][p[i]]
            #print ("из промежуточного города в промежуточный - " + str(a[p[i-1]][p[i]]))
    #print ("из предпоследнего города в последний - " + str(a[p[len(p)-1]][0]))
    sum += a[p[len(p)-1]][0]
    return sum

#print(count([2,1,3])) - для проверки )

#Функция сравнения ans и вновь вычисленного пути из того или иного перебора
def minForAns(count, p):
    global ans
    global res
    if count < ans:
        ans = count
        res = []
        res.append(p)
    elif count == ans:
        res.append(p)
    return ans

def rec(idx):
    global p
    if idx == n-1:
#        print("Путь торговца по городам: " + "0->" + '->'.join(map(str,p)) + "->0")
#        print("     Рассчитанная дальность: " + str(count(p)) + " (минимальная дальность: " + str(minForAns(count(p))) + ")")
        minForAns(count(p), p)
        return
    for i in range(1, n):
        if used[i]:
            continue
        p[idx]=i
        used[i]=True
        rec(idx+1)
        used[i]=False

a = list()
n = int(input())
for i in range(n):
    a.append(list(map(int, input().split())))
    
#Матрица для хранения текущей перестановки
p = [0 for i in range(n-1)]
#Матрица False
used = [False for i in range(n+1)]

res = [] # список для хранения результатов

#Получим длину пути, а чтобы она была максимально большой из возможных - просто просуммируем всё содержимое
ans = 0
for i in range(0, n):
    for j in range(0, n):
        ans += a[i][j]
rec(0)

print(f"Минимальная длина = {ans}, достигнута на", end = ' ')
if len(res) == 1:
    print(f"пути {'->'.join(map(str,res[0]))}")
else:
    print(f"путях:")
    for i in range(len(res)):
        print("0->" + '->'.join(map(str, res[i])) + "->0")

File: алгоритмы перебора/Задача коммивояжера. Перебор с отсечениями/Задача коммивояжера. Перебор с отсечениями/Задача коммивояжера. Перебор с отсечениями.cpp
#include <iostream>
#include <math.h>
#include <vector>

using namespace std;

int** a;            //Массив расстояний    
int* mas;        // Храним Перестановку
bool* used;        //Использование числа
int ans = 100000;//длина пути

int n;
void out()
{
    cout << ans << " :";

    for (int i = 0; i < n; i++)
    {
        cout << mas[i] << " ";
    }
    cout << endl;

}
int min(int a, int b)
{
    if (a < b)return a;
    return b;

}

void rec(int idx, int len)
{

    if (len >= ans) return;

    if (idx == n)//Если заполнили K числа то выходим
    {
        ans = min(ans, len + a[mas[idx - 1]][0]);//добавить длину от посленего города в перестановке до нулевого
        out();//вывод

        return;
    }

    for (int i = 1; i <= n - 1; i++)
    {

        if (used[i]) continue;//Если число занято то следущая итерация цикла

        mas[idx] = i;

        used[i] = true;
        rec(idx + 1, len + a[mas[idx - 1]][i]);//к расстоянию прибаили расстояние от последнего города в перестановке до текущего города i
        used[i] = false;
    }
}

int main()
{


    cin >> n;


    mas = new int[n];

    a = new int* [n];

    for (int i = 0; i < n; i++)
    {
        a[i] = new int[n];

    }

    used = new bool[n];//Храним занятые числа(номера)

    for (int i = 0; i < n; i++)
    {
        used[i] = false;
    }

    for (int i = 0; i < n; i++)
    {
        mas[i] = 0;

    }

    for (int i = 0; i < n; i++)//Заполняем длинами дорог
    {

        for (int j = 0; j < n; j++)
        {
            cin >> a[i][j];

        }
    }

    rec(1, 0);//начинаем с нулевого города (0)-фиксирован остальные значения пробуем
    system("pause");
}